src/main/java/io/cryostat/agent/remote/RecordingsContext.java

```diff
@@ -106,12 +108,23 @@ public void handle(HttpExchange exchange) throws IOException {
                     }
                     break;
                 case "POST":
-                    handleStart(exchange);
+                    handleStartRecordingOrSnapshot(exchange);
                     break;
                 case "PATCH":
                     id = extractId(exchange);
-                    if (id >= 0) {
-                        handleStop(exchange, id);
+                    String request = requestStopOrUpdate(exchange);
+                    if (request == "STOPPED") {
+                        if (id >= 0) {
+                            handleStop(exchange, id);
+                        } else {
+                            exchange.sendResponseHeaders(HttpStatus.SC_BAD_REQUEST, BODY_LENGTH_NONE);
+                        }
+                    } else if (request == "UPDATE") {
```

Comment:
The values for `request` should only be the [`RecordingState`s](https://docs.oracle.com/en/java/javase/11/docs/api/jdk.jfr/jdk/jfr/RecordingState.html). I think you should be able to refactor this so that the Agent accepts the request, parses the body as JSON, iterates over each key in the JSON to determine which property should be set, then validates that the corresponding JSON value is acceptable (ex. for `state` that it can be mapped from the `RecordingState`), and finally performs some update operation on the local recording. The request state check shouldn't be done as a separate top-level check like this, in other words, it should be treated the same as the other properties.

Reply:
This would mean it is possible for a client to issue a request that both renames and stops a recording simultaneously, for example, which would be fine, but your current implementation would not allow.